CONTRIBUTING.md

```diff
@@ -125,6 +125,19 @@ pip3 install pre-commit
 pre-commit install
 ```
 
+## Unit Tests
+To execute all unit tests: (Caution: This takes ~10 minutes on M1 MacOS)
+
+```rippled --unittest```
```

Comment:
fixed in 4dd70570e1eafe6fdec1e1f1ad31590f174a8107

Reply:
When I use parallel threads to run unit tests on a Mac M1, I'm hitting a few failures. But these failures are not observed if I don't use multiple threads. An excerpt of the failures:

```
1> #12 failed: unhandled exception: end of stream [beast.http:1 at /Users/ckeshavabs/.conan/data/boost/1.82.0/_/_/package/a69d410abc9222719bf488aebe7f988537c9471e/include/boost/beast/http/impl/read.hpp:366:59 in function 'std::size_t boost::beast::http::detail::read_some(SyncReadStream &, DynamicBuffer &, basic_parser<isRequest> &, error_code &) [SyncReadStream = boost::asio::basic_stream_socket<boost::asio::ip::tcp>, DynamicBuffer = boost::beast::basic_multi_buffer<std::allocator<char>>, isRequest = false]']
18> #754 failed: apply: Got temINVALID (The transaction is ill-formed.); Expected tesSUCCESS (The transaction was applied. Only final in a validated ledger.)
18> {
   "Account" : "raMjuzZYRZ4yza9DAd5Szt9tz9ZDzD4QzP",
   "Fee" : "10",
   "NFTokenID" : "00000000AE123A8556F3CF91154711376AFB0F894F832B3D9AB9DA6B000000D0",
   "Sequence" : 712,
   "SigningPubKey" : "023F321AFD65B713B5394BBE2C9EE07218985DEDC56C53DE9F89749F9DF56A70ED",
   "TransactionType" : "NFTokenBurn",
   "TxnSignature" : "304402206F8A33ABCA13070D091B5C07594907AFC31FC47653B040EEB0BB9C7E8E97CDC202200F4009CF4922FE150B972E21E27243DADAAC57E597E06EFF2408E9F0E4E728B3"
}

18> #1050 failed: apply: Got tecHAS_OBLIGATIONS (The account cannot be deleted since it has obligations.); Expected tesSUCCESS (The transaction was applied. Only final in a validated ledger.)
18> {
   "Account" : "rG1QQv2nh2gr7RCZ1P8YYcBUKCCN633jCn",
   "Destination" : "rP6P9ypfAmcJpw8SZHNwM4nvZHFXDraQas",
   "Fee" : "2000000",
   "Sequence" : 5,
   "SigningPubKey" : "0388935426E0D08083314842EDFBB2D517BD47699F9A4527318A8E10468C97C052",
   "TransactionType" : "AccountDelete",
   "TxnSignature" : "3045022100A875A2783847BDCBD89431C00B769BDF322DF86ED81E344EBF1B515E8F7B1DDB0220110897E0CF67581DCBECE6931A22B943500A3FFC68AB575FF53D60578FC10950"
}

18> #1051 failed: NFToken_test.cpp(6221)
18> #1056 failed
18> #1059 failed: NFToken_test.cpp(6230)
18> #1062 failed: NFToken_test.cpp(6247)
```

I didn't see any issue for `<32` parallel jobs, but I don't know the exact limits for this behavior.

Reply:
FWIW, I've dug into this a bit, but I've not gotten to the bottom of it.  My belief (unproven) is that the `temINVALID` comes from here:  https://github.com/XRPLF/rippled/blob/develop/src/test/jtx/impl/Env.cpp#L283  Note that the link goes to a _unit test_ file.  So that particular `temINVALID` is not actually coming from the server, it is manufactured by the test framework.

Once that `temINVALID` is handed off by the testing framework all bets are off for the remainder of the test.

I think that the problem is that, when running multithreaded, an RPC response times out waiting for the response.  Again, I've not proven that.  If I'm right then it should be affected by increasing the timeout.  But I did not see an obvious way to increase the timeout, so I've not tested the assumption.

Personally, I always run the unit tests single threaded on my own machine so I don't have to deal with this malfunction.

Reply:
> When I use parallel threads to run unit tests on a Mac M1, I'm hitting a few failures. But these failures are not observed if I don't use multiple threads.

Wow. That's a pretty significant pain in the neck.

> I didn't see any issue for `<32` parallel jobs, but I don't know the exact limits for this behavior.

In that case, what I'd suggest is keep the `--unittest-jobs` change, and add a note specific to Mac M1. Something like: `(Note: Using all the cores on a Mac M1 can cause spurious test failures. The cause is still under investigation. If you observe this problem, try specifying fewer jobs.)`

Reply:
> FWIW, I've dug into this a bit, but I've not gotten to the bottom of it. My belief (unproven) is that the `temINVALID` comes from here: https://github.com/XRPLF/rippled/blob/develop/src/test/jtx/impl/Env.cpp#L283 Note that the link goes to a _unit test_ file. So that particular `temINVALID` is not actually coming from the server, it is manufactured by the test framework.

This has been annoying me for a long time, too. I should do something about it...